### PR TITLE
[release-4.13] OCPBUGS-25611: Avoid panic due to stale ACLs with 1013 priority

### PR DESCRIPTION
**Pull request by arghosh93**
**- What this PR does and why is it needed**

This PR is to take care of following things:

- Remove stale ACLs with priority 1013 and still present in Northbound database.
- Avoid panic due to null pointer. This may occur due to presence of an ACL without a name.

**- Special notes for reviewers**
Jira: https://issues.redhat.com/browse/OCPBUGS-25611

**- How to verify it**
- Create ACL with 1013 priority and no name:
  $ ovn-nbctl --no-leader acl-add clusterRtrPortGroup to-lport 1013 "outport == @ClusterRtrPortGroup && (ip4.mcast || mldv1 || mldv2 || (ip6.dst[120..127] == 0xff && ip6.dst[116] == 1))" allow
 $ ovn-nbctl --no-leader acl-add clusterRtrPortGroup from-lport 1013 "inport == @ClusterRtrPortGroup && (ip4.mcast || mldv1 || mldv2 || (ip6.dst[120..127] == 0xff && ip6.dst[116] == 1))" allow

- Restart ovnkube-master leader 

**- Description for the changelog**
- Remove ACL with priority 1013
- Unit test to validate that ACL with 1013 priority is being removed when syncNetworkPolicies function gets executed
- Avoid panic by checking content of the parameter before trying to access it by address

### CODE REVIEW

File: go-controller/pkg/ovn/policy.go [hunk 1]
```diff
@@ -326,10 +326,28 @@ func (oc *DefaultNetworkController) syncNetworkPolicies(networkPolicies []interf
...
+	if err != nil {
+		return fmt.Errorf("unable delete leftover ACLs from port groups: %v", err)
+	}
+	// End of the hack
 
 	// cleanup port groups based on acl search
 	p := func(item *nbdb.ACL) bool {
-		return item.ExternalIDs[policyACLExtIdKey] != "" || item.ExternalIDs[defaultDenyPolicyTypeACLExtIdKey] != ""
+		return item.ExternalIDs[policyACLExtIdKey] != "" || (item.ExternalIDs[defaultDenyPolicyTypeACLExtIdKey] != "" &&
+			item.Priority != types.DefaultRoutedMcastAllowPriority)
```
Comment (tssurya) [inline go-controller/pkg/ovn/policy.go:350]: little confused with this extra match here, if we are checking for name != nil in L368 do we really need this as well or is it a double precaution?

File: go-controller/pkg/ovn/policy_test.go [hunk 2]
```diff
@@ -1378,6 +1379,47 @@ var _ = ginkgo.Describe("OVN NetworkPolicy Operations", func() {
...
+					"",
+					false,
+					map[string]string{
+						defaultDenyPolicyTypeACLExtIdKey: "Ingress",
+					},
+					nil,
+				)
+				leftOverACL6FromUpgrade.UUID = "ingress-multicast-UUID"
+				testOnlyMulticastPG := libovsdbops.BuildPortGroup(
+					ClusterRtrPortGroupName,
```
Comment (tssurya) [inline go-controller/pkg/ovn/policy_test.go:1416]: do we actually build a new multicast PG in our main logic as well? or do we just use the clusterRtrPG ? checking to see if this is reality or just a test detail.

File: go-controller/pkg/ovn/policy.go [hunk 3]
```diff
@@ -326,10 +326,28 @@ func (oc *DefaultNetworkController) syncNetworkPolicies(networkPolicies []interf
...
+	// https://issues.redhat.com/browse/OCPBUGS-21668
+	// Start of the hack
+	predicate := func(item *nbdb.ACL) bool {
+		return item.Priority == types.DefaultRoutedMcastAllowPriority
+	}
+	staleACLs, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, predicate)
+	if err != nil {
+		return fmt.Errorf("cannot find ACL with priority %d: %v", types.DefaultRoutedMcastAllowPriority, err)
+	}
+	err = libovsdbops.DeleteACLsFromAllPortGroups(oc.nbClient, staleACLs...)
```
Comment (tssurya) [inline go-controller/pkg/ovn/policy.go:341]: should we do this only if len(staleACLs) > 0?

**Review by tssurya — COMMENTED** [review]
a few inline questions..
Overall logic LGTM.
did you also actually do the upgrade test to ensure its all ok @arghosh93 ?

File: go-controller/pkg/ovn/policy.go [hunk 4]
```diff
@@ -326,10 +326,28 @@ func (oc *DefaultNetworkController) syncNetworkPolicies(networkPolicies []interf
 			}
 		}
 	}
+	// Cleanup ACLs with priority 1013
+	// This is a 4.13 specific solution to get rid of a panic caused by presence
+	// of ACL with priority 1013(deprecated) and without a name.
+	// https://issues.redhat.com/browse/OCPBUGS-21668
+	// Start of the hack
```
[the remark this answers is not in the document]
Comment (arghosh93) [reply go-controller/pkg/ovn/policy.go:333]: Noted, I will make it more explicit if I do a force push for any other changes. Thanks for pointing it out !

File: go-controller/pkg/ovn/policy.go (hunk 3 above)
Comment (arghosh93) [reply go-controller/pkg/ovn/policy.go:341]: Yes, it makes sense to check length of  staleACLs before trying to delete . I will do the change and force push. Thanks for pointing it out!

**Comment by arghosh93** [comment]
No @tssurya, I have not tested upgrade to next major version. I will run upgrade and update in case I see any anomaly.

**Comment by arghosh93** [comment]
Haven't seen any issue with upgrade.

[arghosh@arghosh-thinkpadp1gen3 ~]$ oc get clusterversion
NAME      VERSION                                                AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.13.0-0.test-2024-01-29-163427-ci-ln-5yt5mvk-latest   True        False         4m6s    Cluster version is 4.13.0-0.test-2024-01-29-163427-ci-ln-5yt5mvk-latest

[arghosh@arghosh-thinkpadp1gen3 ~]$ oc get clusterversion
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.14.0    True        False         108s    Cluster version is 4.14.0

File: go-controller/pkg/ovn/policy_test.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy_test.go:1416]: ack consider this resolved

File: go-controller/pkg/ovn/policy.go (hunk 1 above)
Comment (tssurya) [reply go-controller/pkg/ovn/policy.go:352]: sg, resolve.

**Review by tssurya — APPROVED** [review]
let CI run before I ask for backport risk label.

**Review by tssurya — APPROVED** [review]
/lgtm

**Comment by tssurya** [comment]
```
: [sig-api-machinery][Feature:ClusterResourceQuota] Cluster resource quota should control resource limits across namespaces [apigroup:quota.openshift.io][apigroup:image.openshift.io] [Suite:openshift/conformance/parallel] 
```
unrelated failure

**Pull request merged**